### ISSUE TITLE
Error printing to logs from Godot is not rate limited

### ISSUE THREAD

**Issue by hhyyrylainen** [godotengine/godot#45181]
**Godot version:**
3.2.3.stable.mono (also previous versions at least as far back as 3.2.1)

**OS/device including version:**
GLES3, both Windows and Linux

**Issue description:**
When Godot engine encounters an error, it prints it to logs. It's good that all error text is printed (and ignores normal print limit per second), however I've experienced this a few times and gotten reports from players, that in some cases when Godot really locks up on some error, it will infinitely spam an error message as fast as possible. One user reported that before they managed to kill the Godot process, the log file was over 2 gigabytes large. I can't guarantee this, but I've also received 3 reports total that my game, that uses Godot, has caused the user's computer to totally lockup, and in 2 cases rebooting and trying to run the game again didn't work (the game files had become corrupted). So I very strongly think that there needs to be some limit in all Godot error printing to stop gigabytes of logs being generated that freeze computers.

This issue is related to: https://github.com/godotengine/godot/issues/37652 however this is not the same, as I'm saying that all of the error printing code should have an inbuilt fail-safe to kill the process before ridiculous amounts of repeating logs are printed.

Game using Godot froze / locked up (presumably) in a physics related bug. Instead of the game properly crashing and process exiting, instead the process just stayed open printing the same repeating error in log files that ended up taking gigabytes of space. I expected that the engine would have some kind of fail-safe to just give up on writing logs at some point.

I tried to find an issue about this, but I only found specific cases where infinite errors are printed, and not a general report about the problem of really big log files.

**Steps to reproduce:**
Enable logging to a file, then hit a very serious bug in godot, for example https://github.com/godotengine/godot/issues/37652 and wait some time while Godot is frozen to generate really big log files.

**Comment by Calinou** [comment]
If this is about Mono logs, this is a duplicate of https://github.com/godotengine/godot/issues/21895.

**Comment by akien-mga** [comment]
> If this is about Mono logs (which are separate from engine logs), this is a duplicate of #21895.

That's not a Mono specific issue, though Mono does make the issue more critical as it tends to be quite spammy.

> As for the issue mentioned here, I'm not sure if we can do much about it. If we disable writing to a file once the log has exceeded a certain size, we risk losing out on important information especially if the process does end up exiting at some point in time.

At that point I don't think it matters, if your logs are 2 GB most users wouldn't even be able to extract the tail of the log which might possibly contain useful info (and most likely wouldn't).

**Comment by hhyyrylainen** [comment]
This is not about mono logs. It's these logs that I mean, that are from the game and from Godot engine:
![Kuvakaappaus - 2021-01-14 16-07-37](https://user-images.githubusercontent.com/3796411/104601292-b9ed3580-5682-11eb-81f8-7ab4d3ff5c39.png)

I don't usually look at the mono logs and I haven't instructed any players to look at them either.

So far each time a single message repeats forever. I don't really have a good suggestion on where to draw the line, but detection on when to stop, could rely on detecting that the last message just repeats for megabytes worth of logs.

> At that point I don't think it matters, if your logs are 2 GB most users wouldn't even be able to extract the tail of the log which might possibly contain useful info (and most likely wouldn't).

Exactly, some users even asked me advice on what tool to use to even open a log that is so large so that they could post the first few megabytes for me to seen.

**Comment by Calinou** [comment]
I'm not sure how to fix this.

- Limiting to a given file size (such as 20 MB) isn't a good idea for servers or persistent processes since these can run for months with no interruption.
- Limiting using a timer isn't a bad idea, but I don't think we can use Timer nodes or stuff like that in the rotating file logger since it's a core API that should work without a scene tree.
- Detecting endless message loops is difficult as you may have more than 1 message being printed in a quick succession (such as a chain of 3-4 messages).

Any ideas?

**Comment by akien-mga** [comment]
Fixing this would also be necessary before we can enable saving editor logs to files too, and not just project logs. Just like Mono, the editor can be pretty spammy as it's easy to trigger errors in a process callback while developing.
